src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts

```diff
@@ -601,4 +601,72 @@ describe('TableComponent', () => {
       expect(component.useCustomClass('https://secure.it')).toBe('btn secure');
     });
   });
+
+  describe('test expand and collapse feature', () => {
```

Comment:
I looked through your tests and did a minor refactoring, if you like it use it .

```diff
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts
index b5fa4a35c9..3a12e5d789 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/table/table.component.spec.ts
@@ -604,65 +604,64 @@ describe('TableComponent', () => {
 
   describe('test expand and collapse feature', () => {
     beforeEach(() => {
+      spyOn(component.setExpandedRow, 'emit');
+      // Table row detail mock
+      component.table = {
+        rowDetail: { collapseAllRows: jest.fn(), toggleExpandRow: jest.fn() }
+      } as any;
+      // Setup table
       component.identifier = 'a';
       component.data = createFakeData(10);
+      // Select item
       component.expanded = _.clone(component.data[1]);
     });
 
-    it('should update expaneded on refresh - "always"', () => {
-      spyOn(component.setExpandedRow, 'emit');
-      component.updateExpandedOnRefresh = 'always';
-
-      component.data[1].b = 'test';
-      component.updateExpanded();
-      expect(component.expanded.b).toBe('test');
-      expect(component.setExpandedRow.emit).toHaveBeenCalled();
-    });
+    describe('update expanded on refresh', () => {
+      const updateExpendedOnState = (state: 'always' | 'never' | 'onChange') => {
+        component.updateExpandedOnRefresh = state;
+        component.updateExpanded();
+      }
 
-    it('should update expaneded on refresh - "onChange" with changes', () => {
-      spyOn(component.setExpandedRow, 'emit');
-      component.updateExpandedOnRefresh = 'onChange';
-
-      component.data[1].b = 'test';
-      component.updateExpanded();
-      expect(component.expanded.b).toBe('test');
-      expect(component.setExpandedRow.emit).toHaveBeenCalled();
-    });
-
-    it('should update expaneded on refresh - "onChange" without changes', () => {
-      spyOn(component.setExpandedRow, 'emit');
-      component.updateExpandedOnRefresh = 'onChange';
+      beforeEach(() => {
+        // Mock change
+        component.data[1].b = 'test';
+      })
+
+      it('refreshes "always"', () => {
+        updateExpendedOnState('always')
+        expect(component.expanded.b).toBe('test');
+        expect(component.setExpandedRow.emit).toHaveBeenCalled();
+      });
 
-      component.updateExpanded();
-      expect(component.expanded.b).toBe(10);
-      expect(component.setExpandedRow.emit).not.toHaveBeenCalled();
-    });
+      it('refreshes "onChange"', () => {
+        updateExpendedOnState('onChange')
+        expect(component.expanded.b).toBe('test');
+        expect(component.setExpandedRow.emit).toHaveBeenCalled();
+      });
 
-    it('should update expaneded on refresh - "never"', () => {
-      spyOn(component.setExpandedRow, 'emit');
-      component.updateExpandedOnRefresh = 'never';
+      it('does not refresh "onChange" if data is equal', () => {
+        component.data[1].b = 10; // Reverts change
+        updateExpendedOnState('onChange')
+        expect(component.expanded.b).toBe(10);
+        expect(component.setExpandedRow.emit).not.toHaveBeenCalled();
+      });
 
-      component.data[1].b = 'test';
-      component.updateExpanded();
-      expect(component.expanded.b).toBe(10);
-      expect(component.setExpandedRow.emit).not.toHaveBeenCalled();
-    });
+      it('"never" refreshes', () => {
+        updateExpendedOnState('never')
+        expect(component.expanded.b).toBe(10);
+        expect(component.setExpandedRow.emit).not.toHaveBeenCalled();
+      });
+    })
 
     it('should open the table details and close other expanded rows', () => {
-      spyOn(component.setExpandedRow, 'emit');
-      component.table = {
-        rowDetail: { collapseAllRows: jest.fn(), toggleExpandRow: jest.fn() }
-      } as any;
       component.toggleExpandRow(component.expanded, false);
-      expect(component.expanded).toBe(component.expanded);
+      expect(component.expanded).toEqual({ a: 1, b: 10, c: true });
       expect(component.table.rowDetail.collapseAllRows).toHaveBeenCalled();
       expect(component.setExpandedRow.emit).toHaveBeenCalledWith(component.expanded);
       expect(component.table.rowDetail.toggleExpandRow).toHaveBeenCalled();
     });
 
     it('should close the current table details expansion', () => {
-      spyOn(component.setExpandedRow, 'emit');
-      component.table = { rowDetail: { toggleExpandRow: jest.fn() } } as any;
       component.toggleExpandRow(component.expanded, true);
       expect(component.expanded).toBeUndefined();
       expect(component.setExpandedRow.emit).toHaveBeenCalledWith(undefined);
```